WordPress/openverse-catalog
# Add DAG to report reported media pending review

## Fixes


Fixes #354

## Description


Adds a `report_pending_reported_images` DAG that runs weekly and for each media type:
* checks to see if there are any reported records in the `pending_review` status, meaning they require manual review through the Django Admin
* reports the number of _distinct_ records needing review to Slack, as well as a breakdown of the reasons for reporting (dmca, mature, other) and links to the admin page for that media

<img width="511" alt="Screen Shot 2022-05-20 at 3 43 07 PM" src="https://user-images.githubusercontent.com/63313398/169621412-d6fb524b-bfd0-4619-9061-c169d04e6b46.png">


## Additional Notes
* I verified that if a single image has multiple reports in `pending_review` associated with it, when you update _any_ of these records _all_ the other records associated to that image are updated as well. (This is why I think it's best to report _distinct_ images requiring review)
* When new media types are added, this will require you to add the name of the reports table in the configuration.

## Testing Instructions

This requires a connection to the API DB (rather than the Catalog), so you'll need to create the connection env variables. You can use the values set in `env.template` in this PR. Make sure you also have the API running locally. To test the slack messages you can create the Airflow variable `slack_message_override` and set it to `true`.

Run the `report_pending_reported_media` DAG locally. If you're running sample data, there should be no reports and you should see the `No records to report` message.

To create sample data to test the report count, you can report some images and audio (locally!) through the API. I went to `localhost:8000/v1/images` and grabbed the `id` for a couple of the listed images. For each I then navigated to `localhost:8000/v1/<id>/report` and POSTed a report. You can verify that these went through by checking out `http://localhost:8000/admin/api/imagereport/?` Likewise for Audio. Report a mix of different reasons (some 'mature', some 'other', etc).

Run the DAG again and make sure the correct numbers are reported. I also played around with manually reviewing some of the media reports (updating their status to something other than `pending_review` in the Admin) and verifying they were no longer included in the count; and also reporting the same records multiple times and verifying that it is only counted once in the output.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
